zkevm-circuits/src/blobkzg_circuit/src/main.rs

```diff
@@ -0,0 +1,585 @@
+            evals: [F::default(); 4098],
+            wis: [F::default(); 4096],
+            xns: [F::default(); 257],
+            x2is: [F::default(); 256],
+            nbits: [F::default(); 256],
+        }
+    }
+}
+impl<W: PrimeField> WitnessValues<W> {
+    /// COMPUTATION OF PRIVATE WITNESS DATA
+    /// we have two main long computations:
+    /// a. the eval step
+    /// b. the xns step
+    ///
+    /// for the eval step we apply the main formula
+    /// to the datapoints, requiring the intermediate
+    /// rots of unity (wis) to be computed, as well as
+    /// the final exponentiation x^n (xns) to be computed.
+    ///
+    /// for the xns step we compute the square-and-multiply
+    /// algorithm, putting together all the possible powers
+    /// (x2is) to be computed and all the exponent's bits
+    /// (nbits) to be computed such that only the necessary powers
+    /// of x are multiplied together.
+    fn new(
+        InputValues {
+            datapoints,
+            x,
+            w,
+            n,
+        }: &InputValues<W>,
+    ) -> WitnessValues<W> {
+        let x2is: [W; 256] = {
+            let mut x2is = [W::ZERO; 256];
+            let mut state = *x;
+            for x in &mut x2is {
+                *x = state;
+                state *= state;
+            }
+            x2is
+        };
+        let nbits: [W; 256] = {
+            let mut nbits = [W::ZERO; 256];
+            let nbits_bool: [bool; 256] = {
+                let mut nbits_bool = [false; 256];
+                let repr: [u8; 32] = n.to_repr().as_ref().try_into().unwrap();
+                for (j, byte) in repr.iter().enumerate() {
+                    for i in 0..8 {
+                        nbits_bool[j * 8 + i] = ((byte >> i) & 1) == 1;
+                    }
+                }
+                nbits_bool
+            };
+
+            for (x, b) in nbits.iter_mut().zip(nbits_bool.iter()) {
+                if *b {
+                    *x = W::ONE
+                }
+            }
+            nbits
+        };
+        let xns: [W; 257] = {
+            let mut xns = [W::ZERO; 257];
+            let mut state = W::ONE;
+            for (x, (x2i, nbit)) in xns.iter_mut().zip(x2is.iter().zip(nbits.iter())) {
+                *x = state;
+                state *= *x2i * *nbit + (W::ONE - *nbit);
+            }
+            xns[xns.len() - 1] = state;
+            xns
+        };
+        let wis: [W; 4096] = {
+            let mut wis = [W::ZERO; 4096];
+            let mut state = *w;
+            for x in &mut wis {
+                *x = state;
+                state *= w;
+            }
+            wis
+        };
+        let evals: [W; 4098] = {
+            let mut evals = [W::ZERO; 4098];
+            let mut state = W::ZERO;
+            for (eval, (di, wi)) in evals.iter_mut().zip(datapoints.iter().zip(wis.iter())) {
+                *eval = state;
+                state += *di * *wi * (*x - *wi).invert().unwrap();
```

Comment:
Couldn't `x == wi` here? given it is an arbitrary input. that way, we should probably return a `Result::Err` instead of panic, as this might become an attack vector